bst.py

```diff
@@ -0,0 +1,127 @@
+from tree_node import Node
+
+
+class BST:
+    def __init__(self) -> None:
+        self.root = None
+    
+    def add_node(self, element):
+        if not self.root:
+            self.root = Node(element)
+            return True
+        
+        temp = self.root
+        while temp:
+            if element < temp.value:
+                if not temp.left:
+                    temp.left = Node(element)
+                    return True
+                temp = temp.left
+            else:
+                if not temp.right:
+                    temp.right = Node(element)
+                    return True
+                temp = temp.right
+    
+    def pre_order_traversal(self, node):
+        temp = node
+        if temp:
+            print(temp.value)
+            self.pre_order_traversal(temp.left)
+            self.pre_order_traversal(temp.right)
+        
+    def post_order_traversal(self, node):
+        temp = node
+        if temp:
+            self.pre_order_traversal(temp.left)
+            self.pre_order_traversal(temp.right)
+            print(temp.value)
+
+    def in_order_traversal(self, node):
+        temp = node
+        if temp:
+            self.pre_order_traversal(temp.left)
+            print(temp.value)
+            self.pre_order_traversal(temp.right)
+
+    def search_node(self, element):
+        if not self.root:
+            return None
+        temp = self.root
+        while temp:
+            if temp.value == element:
+                return temp
+            elif element < temp.value:
+                temp = temp.left
+            else:
+                temp = temp.right
+
+        return None
+    
+    def delete_node(self, element):
+        node = self.search_node(element)
+        if not node:
+            return None
+        if not node.right and node.left:
+            node.value = None
+        elif not node.right and node.left:
+            pass
+        else:
+            pass
```

Comment:
So, when we delete the BST... We need to maintain its properly of  Root's left child holds lesser value and right child holds greater value. There are 2 ways to do the removal operation. 
```
Step1 Find the root to delete
Step2 Find its leftmost RIGHT leaf node OR Find its rightmost LEFT leaf node
Step3 Swap them
Step4 Delete that leaf node
```
Principle behind it is, the leftmost RIGHT leaf node OR rightmost LEFT leaf node fulfil the property of BST. Draw some valid BSTs and try it!

Reply:
As I read we may need to delete any of the node in the BST not only root node. For that logic will be different.